game/kernel/asm_funcs_arm64.s

```diff
@@ -0,0 +1,262 @@
+  ; arg 1 (RSI in x86)
+  ; arg 0 (RDI in x86)
+  stp x7, x6, [sp, #-16]!
+  stp x5, x4, [sp, #-16]!
+  stp x3, x2, [sp, #-16]!
+  stp x1, x0, [sp, #-16]!
+
+  ; set first argument
+  mov x19, sp
+  ; call function
+  blr x8
+  ; restore arguments
+  ldp x0, x1, [sp], #16
+  ldp x2, x3, [sp], #16
+  ldp x4, x5, [sp], #16
+  ldp x6, x7, [sp], #16
+
+  ldp x8, x9, [sp], #32
+  ldp x10, x11, [sp], #32
+  ldp x12, x13, [sp], #32
+  ldp x14, x15, [sp], #32
+
+  ; return!
+  ret
+
+;; Call c++ code through mips2c.
+;; GOAL will call a dynamically generated trampoline.
+;; The trampoline will have pushed the exec function and stack offset onto the stack
+.global _mips2c_call_arm64
+.align 4
+_mips2c_call_arm64:
+  ;; TODO - this is really weird using half an XMM, this makes the arm assembly
+  ;; more difficult - this probably isn't required for arm?
+  ;; grab the address to call and put it in xmm0
+  ;; TODO - this stack pointer manipulation might be a problem for ARM64 which requires 16byte alignment
+  ;; sub sp, 8
+  ldr q0, [sp, #+16]
+  ;; grab the stack offset
+  ldr x0, [sp, #+8]
+
+  ;; first, save quadword registers
+  stp q15, q14, [sp, #-32]!
+  stp q13, q12, [sp, #-32]!
+  stp q11, q10, [sp, #-32]!
+  stp q9, q8, [sp, #-32]!
+
+  ; NOTE - in x86 the 2 special registers are saved (R10 and R11)
+  ; we don't need to do that in ARM64, there are plenty of registers to work with
+
+  ;; oof
+  sub sp, sp, 1280
+  str x0, [sp, #+64] ; arg 0 (RDI in x86) and 
+  str x1, [sp, #+80] ; arg 1 (RSI in x86)
+  str x2, [sp, #+96] ; arg 2 (RDX in x86) and arg 3 (RCX in x86)
+  str x3, [sp, #+112] ; arg 2 (RDX in x86) and arg 3 (RCX in x86)
+  str x4, [sp, #+128] ; arg 4 (R8 in x86) and arg 5 (R8 in x86)
+  str x5, [sp, #+144] ; arg 4 (R8 in x86) and arg 5 (R8 in x86)
+  str x6, [sp, #+160] ; arg 6 (R10 in x86) and arg 7 (R11 in x86)
+  str x7, [sp, #+176] ; arg 6 (R10 in x86) and arg 7 (R11 in x86)
+  str x20, [sp, #+352] ;; s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
+  str x21, [sp, #+368] ;; s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
+
+  mov x0, sp ; move the stack pointer to arg 0
+  sub x0, x0, x22 ; R15 is a "special" offset TODO - whats special about it?
+  str x0, [sp, #+464] ;; mip2c code's MIPS stack
+
+  mov x0, sp ;; move the stack pointer to the new position
+
+  sub sp, sp, x8 ;; allocate space on the stack for GOAL fake stack
+  stp x8, x8, [sp, #-16]! ;; and remember this so we can find our way back
+
+  ;; TODO - this used to be a movq rax, xmm0
+  ;; TODO - not sure why an `xmm` was used because that movq only uses the lower 64bits anyway
+  mov x0, v0.d[0] ; represents the lower 64 bits of q0
+  blr x8 ;; call!
+
+  ;; unallocate
+  ldp x8, x8, [sp], #16
+  add sp, sp, x8
+
+  ldr x8, [sp, #+32]
+
+  add sp, sp, 1280 ; reset the stackpointer back
+
+  ldp x8, x9, [sp], #32
+  ldp x10, x11, [sp], #32
+  ldp x12, x13, [sp], #32
+  ldp x14, x15, [sp], #32
+
+  add sp, sp, 24 ;; 16 for the stuff pushed by trampoline
```

Comment:
x86 also needs 16-byte aligned stacks, but `call`/`ret` on x86 push/pop 8 byte values (but this is not the case on arm). All these offsets and stuff assume the 8-bytes are added and might need to be redone for arm (`blr` stores the return address in a register instead of pushing).

Reply:
Somewhat related -- what is the reason why `rax` is pushed twice here? https://github.com/open-goal/jak-project/blob/33858cbae6ab2c68958f4ac5f56619a249e3433c/game/kernel/asm_funcs.asm#L151-L160

Reply:
This was just me being dumb. The second is "needed" to align the stack properly to 16-bytes. But this could be replaced with `sub rsp, 8` and `add rsp, 8`